File: source/FAST/Visualization/Shortcuts.cpp
#include "Shortcuts.hpp"
#include "SlicerWindow.hpp"
#include <FAST/Visualization/ImageRenderer/ImageRenderer.hpp>
#include <FAST/Visualization/ImagePyramidRenderer/ImagePyramidRenderer.hpp>
#include <FAST/Visualization/SegmentationRenderer/SegmentationRenderer.hpp>
#include <FAST/Visualization/SimpleWindow.hpp>
#include <FAST/Visualization/RenderToImage/RenderToImage.hpp>
#include <FAST/Data/ImagePyramid.hpp>
#include <FAST/Visualization/VertexRenderer/VertexRenderer.hpp>
#include <FAST/Visualization/VolumeRenderer/AlphaBlendingVolumeRenderer.hpp>
#include <FAST/Visualization/VolumeRenderer/MaximumIntensityProjection.hpp>

namespace fast {

template <class T>
bool hasValue(T type) {
    return !std::holds_alternative<std::monostate>(type);
}

template <class T>
T getValueOrDefault(std::optional<T> variable, T defaultValue) {
    if(variable.has_value()) {
        return variable.value();
    } else {
        return defaultValue;
    }
}

std::variant<std::monostate, Window::pointer, Image::pointer> display2D(Display2DArgs args) {
    if(!hasValue(args.image) &&
       !hasValue(args.segmentation) &&
       !hasValue(args.imagePyramid) &&
       !hasValue(args.vertices) &&
       !hasValue(args.lines)
       ) {
        throw Exception("No data given to display2D");
    }

    std::vector<Renderer::pointer> renderers;
    if(hasValue(args.image)) {
        auto renderer = ImageRenderer::create(
                getValueOrDefault(args.intensityLevel, -1.0f),
                getValueOrDefault(args.intensityWindow, -1.0f)
        );
        if(std::holds_alternative<Image::pointer>(args.image)) {
            renderer->connect(std::get<Image::pointer>(args.image));
        } else {
            renderer->connect(std::get<ProcessObject::pointer>(args.image));
        }
        renderers.push_back(renderer);
    }

    if(hasValue(args.imagePyramid)) {
        auto renderer = ImagePyramidRenderer::create();
        if(std::holds_alternative<ImagePyramid::pointer>(args.imagePyramid)) {
            renderer->connect(std::get<ImagePyramid::pointer>(args.imagePyramid));
        } else {
            renderer->connect(std::get<ProcessObject::pointer>(args.imagePyramid));
        }
        renderers.push_back(renderer);
    }

    if(hasValue(args.segmentation)) {
        auto renderer = SegmentationRenderer::create(
                args.segmentationColors,
                args.segmentationOpacity,
                args.segmentationBorderOpacity,
                args.segmentationBorderRadius
        );
        if(std::holds_alternative<Image::pointer>(args.segmentation)) {
            renderer->connect(std::get<Image::pointer>(args.segmentation));
        } else if(std::holds_alternative<ImagePyramid::pointer>(args.segmentation)) {
            renderer->connect(std::get<ImagePyramid::pointer>(args.segmentation));
        } else {
            renderer->connect(std::get<ProcessObject::pointer>(args.segmentation));
        }
        renderers.push_back(renderer);
    }

    if(hasValue(args.vertices)) {
        auto renderer = VertexRenderer::create(args.vertexSize, args.vertexSizeIsInPixels, args.vertexMinSize, args.vertexColor, args.vertexLabelColors, args.vertexOpacity);
        if(std::holds_alternative<Mesh::pointer>(args.vertices)) {
            renderer->connect(std::get<Mesh::pointer>(args.vertices));
        } else {
            renderer->connect(std::get<ProcessObject::pointer>(args.vertices));
        }
        renderers.push_back(renderer);
    }

    if(hasValue(args.lines)) {
        auto renderer = LineRenderer::create(args.lineColor, args.lineWidth);
        if(std::holds_alternative<Mesh::pointer>(args.lines)) {
            renderer->connect(std::get<Mesh::pointer>(args.lines));
        } else {
            renderer->connect(std::get<ProcessObject::pointer>(args.lines));
        }
        renderers.push_back(renderer);
    }

    if(args.renderToImage) {
        int width = args.width;
        if(width == 0)
            width = 1024;
        int height = args.height;
        if(height == 0)
            height = -1;
        auto renderToImage = RenderToImage::create(args.bgcolor, width, height)
                ->connect(renderers);
        return renderToImage->runAndGetOutputData<Image>();
    } else {
        auto window = SimpleWindow2D::create(args.bgcolor, args.width, args.height)
                ->connect(renderers);
        if(args.fullscreen)
            window->enableFullscreen();
        if(args.maximize)
            window->enableMaximized();
        if(!args.title.empty())
            window->setTitle(args.title);
        if(hasValue(args.widgets)) {
            if(std::holds_alternative<std::vector<QWidget*>>(args.widgets)) {
                window->connect(std::get<std::vector<QWidget*>>(args.widgets));
            } else {
                for(const auto& item : std::get<std::map<WidgetPosition, std::vector<QWidget*>>>(args.widgets)) {
                    window->connect(item.second, item.first);
                }
            }
        }
        if(args.timeout.has_value())
            window->setTimeout(args.timeout.value());
        if(args.returnWindow) {
            return window;
        } else {
            window->run();
            return std::monostate{};
        }
    }
}

std::variant<std::monostate, Window::pointer> display3D(Display3DArgs args) {
    if(!hasValue(args.image) &&
       !hasValue(args.segmentation)
            ) {
        throw Exception("No data given to display3D");
    }

    std::shared_ptr<Window> window;
    if(args.displayType == DisplayType::SLICER) {
        auto slicerWindow = SlicerWindow::create(args.bgcolor, args.width, args.height);
        if(args.fullscreen)
            slicerWindow->enableFullscreen();
        if(args.maximize)
            slicerWindow->enableMaximized();
        if(!args.title.empty())
            slicerWindow->setTitle(args.title);
        if(hasValue(args.image)) {
            if(std::holds_alternative<Image::pointer>(args.image)) {
                slicerWindow->connectImage(std::get<Image::pointer>(args.image), getValueOrDefault(args.intensityLevel, -1.0f),
                                      getValueOrDefault(args.intensityWindow, -1.0f));
            } else {
                slicerWindow->connectImage(std::get<ProcessObject::pointer>(args.image), getValueOrDefault(args.intensityLevel, -1.0f),
                                           getValueOrDefault(args.intensityWindow, -1.0f));
            }
        }
        if(hasValue(args.segmentation)) {
            if(std::holds_alternative<Image::pointer>(args.segmentation)) {
                slicerWindow->connectSegmentation(std::get<Image::pointer>(args.segmentation), args.segmentationColors, args.segmentationOpacity, args.segmentationBorderOpacity, args.segmentationBorderRadius);
            } else {
                slicerWindow->connectSegmentation(std::get<ProcessObject::pointer>(args.segmentation), args.segmentationColors, args.segmentationOpacity, args.segmentationBorderOpacity, args.segmentationBorderRadius);
            }
        }
        window = slicerWindow;
    } else {
        if(!hasValue(args.image))
            throw Exception("display3D must have image for volume rendering");
        auto simpleWindow = SimpleWindow3D::create(args.bgcolor, args.width, args.height);
        if(args.fullscreen)
            simpleWindow->enableFullscreen();
        if(args.maximize)
            simpleWindow->enableMaximized();
        if(!args.title.empty())
            simpleWindow->setTitle(args.title);
        if(args.displayType == DisplayType::ALPHA_BLENDING) {
            auto renderer = AlphaBlendingVolumeRenderer::create(args.transferFunction);
            if(std::holds_alternative<Image::pointer>(args.image)) {
                renderer->connect(std::get<Image::pointer>(args.image));
            } else {
                renderer->connect(std::get<ProcessObject::pointer>(args.image));
            }
            simpleWindow->addRenderer(renderer);
        } else if(args.displayType == DisplayType::MAXIMUM_INTENSITY_PROJECTION) {
            auto renderer = MaximumIntensityProjection::create();
            if(std::holds_alternative<Image::pointer>(args.image)) {
                renderer->connect(std::get<Image::pointer>(args.image));
            } else {
                renderer->connect(std::get<ProcessObject::pointer>(args.image));
            }
            simpleWindow->addRenderer(renderer);
        } else {
            throw Exception("Invalid display type");
        }
        window = simpleWindow;
    }

    if(hasValue(args.widgets)) {
        if(std::holds_alternative<std::vector<QWidget*>>(args.widgets)) {
            window->connect(std::get<std::vector<QWidget*>>(args.widgets));
        } else {
            for(const auto& item : std::get<std::map<WidgetPosition, std::vector<QWidget*>>>(args.widgets)) {
                window->connect(item.second, item.first);
            }
        }
    }
    if(args.timeout.has_value())
        window->setTimeout(args.timeout.value());
    if(args.returnWindow) {
        return window;
    } else {
        window->run();
        return std::monostate{};
    }
}

}

File: source/FAST/Python/Shortcuts.py
"""
A few shortcut functions
"""
from typing import Dict, Union
import fast

def _set_default_shorcut_value(x, default):
    return default if x is None else x

def display2D(
            image:Union[Image, ProcessObject]=None,
            segmentation:Union[Image, ImagePyramid, ProcessObject]=None,
            imagePyramid:Union[ImagePyramid, ProcessObject]=None,
            lines:Union[Mesh, ProcessObject]=None,
            vertices: Union[Mesh, ProcessObject]=None,
            intensityLevel:float=None,
            intensityWindow:float=None,
            segmentationColors:Dict[int, Color]=None,
            segmentationOpacity:float=0.5,
            segmentationBorderOpacity:float=None,
            segmentationBorderRadius:int=1,
            lineWidth:float=1,
            lineColor:Color=Color.Green(),
            vertexSize:float=10.0,
            vertexSizeIsInPixels:bool=True,
            vertexMinSize:int=1,
            vertexColor:Color=Color.Null(),
            vertexLabelColors:Dict[int, Color]=None,
            vertexOpacity:float=1.0,
            bgcolor:Color=Color.White(),
            width:int=None,
            height:int=None,
            timeout:int=None,
            renderToImage:bool=False,
            returnWindow:bool=False,
            widgets:Union[List, Dict[int, List]]=None,
            title:str='',
            fullscreen:bool=False,
            maximize:bool=False,
        ):
    """
    Shortcut for displaying image, segmentation and meshes using SimpleWindow2D

    :param image: Image to display
    :param segmentation: Segmentation to display
    :param imagePyramid: ImagePyramid to display
    :param lines: Lines to display
    :param vertices: Vertices to display
    :param intensityLevel: Intensity level for image rendering
    :param intensityWindow: Intensity window for image rendering
    :param segmentationColors: Colors to use for segmentation
    :param segmentationOpacity: Opacity of segmentation
    :param segmentationBorderOpacity: Border opacity of segmentation
    :param segmentationBorderRadius: Size of segmentation border
    :param lineWidth: Width of line
    :param lineColor: Color of line
    :param vertexSize Vertex point size (can be in pixels or millimeters, see sizeIsInPixels parameter)
    :param vertexSizeIsInPixels Whether size is given in pixels or millimeters
    :param vertexMinSize Minimum size in pixels, used when sizeInPixels = false
    :param vertexColor Override color stored for each vertex
    :param vertexLabelColors Set vertex color per label
    :param vertexOpacity Opacity of vertices: 1 = no transparency, 0 = fully transparent
    :param bgcolor: Background color
    :param width: Width of window
    :param height: Height of window
    :param timeout: If set to a number, the window will auto close after this many milliseconds
    :param renderToImage: Use RenderToImage instead of SimpleWindow and return the resulting image
    :param returnWindow: Whether to return the window object, or to run it
    :param widgets: Widgets to connect to the window
    :param title: Title to set to window, if renderToImage is false.
    :param fullscreen: Enable fullscreen window, if renderToImage is false.
    :param maximize: Maximize window, if renderToImage is false.
    :return: window if returnWindow is set to True, else None
    """

    if image is None and imagePyramid is None and segmentation is None and lines is None and vertices is None:
        raise ValueError('No data was given to display2D')

    width = _set_default_shorcut_value(width, 0)
    height = _set_default_shorcut_value(height, 0)
    intensityLevel = _set_default_shorcut_value(intensityLevel, -1)
    intensityWindow = _set_default_shorcut_value(intensityWindow, -1)
    segmentationColors = _set_default_shorcut_value(segmentationColors, LabelColors())
    segmentationBorderOpacity = _set_default_shorcut_value(segmentationBorderOpacity, -1)

    renderers = []

    if image is not None:
        renderer = ImageRenderer.create(
            level=intensityLevel,
            window=intensityWindow
        ).connect(image)
        renderers.append(renderer)

    if imagePyramid is not None:
        renderer = ImagePyramidRenderer.create(
        ).connect(imagePyramid)
        renderers.append(renderer)

    if segmentation is not None:
        renderer = SegmentationRenderer.create(
            colors=segmentationColors,
            opacity=segmentationOpacity,
            borderOpacity=segmentationBorderOpacity,
            borderRadius=segmentationBorderRadius
        ).connect(segmentation)
        renderers.append(renderer)

    if lines is not None:
        renderer = LineRenderer.create(
            lineWidth=lineWidth,
            color=lineColor
        ).connect(lines)
        renderers.append(renderer)

    if vertices is not None:
        renderer = VertexRenderer.create(
            size=vertexSize,
            sizeIsInPixels=vertexSizeIsInPixels,
            minSize=vertexMinSize,
            color=vertexColor,
            opacity=vertexOpacity,
            labelColors=vertexLabelColors
        ).connect(vertices)
        renderers.append(renderer)

    if renderToImage:
        render = RenderToImage.create(
            bgcolor=bgcolor,
            width=width,
            height=height
        ).connect(renderers)
        return render.runAndGetOutputData()
    else:
        window = SimpleWindow2D.create(
            bgcolor=bgcolor,
            width=width,
            height=height
        ).connect(renderers)
        if title: window.setTitle(title)
        if fullscreen: window.enableFullscreen()
        if maximize: window.enableMaximized()
        if widgets:
            if isinstance(widgets, dict):
                for pos, widget_list in widgets.items():
                    window.connect(widget_list, pos)
            else:
                window.connect(widgets)
        if timeout:
            window.setTimeout(timeout)
        if returnWindow:
            return window
        else:
            window.run()


from enum import Enum
class DisplayType(Enum):
    SLICER = 1
    ALPHA_BLENDING = 2
    MAXIMUM_INTENSITY_PROJECTION = 3

def display3D(
        image:Union[Image, ProcessObject]=None,
        segmentation:Union[Image, ProcessObject]=None,
        intensityLevel:float=None,
        intensityWindow:float=None,
        segmentationColors:Dict[int, Color]=None,
        segmentationOpacity:float=0.5,
        segmentationBorderOpacity:float=None,
        segmentationBorderRadius:int=1,
        transferFunction:TransferFunction=None,
        displayType:DisplayType=DisplayType.SLICER,
        bgcolor:Color=Color.White(),
        width:int=None,
        height:int=None,
        timeout:int=None,
        returnWindow:bool=False,
        widgets:Union[List, Dict[int, List]]=None,
        title:str='',
        fullscreen:bool=False,
        maximize:bool=False,
):
    """
    Shortcut for displaying image, segmentation and meshes using SimpleWindow2D

    TODO:
        * Surface extraction or threshold volume rendering?
        * Mesh support: Vertices, lines and triangles

    :param image: Image to display
    :param segmentation: Segmentation to display
    :param bgcolor: Background color
    :param width: Width of window
    :param height: Height of window
    :param timeout: If set to a number, the window will auto close after this many milliseconds
    :param returnWindow: Whether to return the window object, or to run it
    :param widgets: Widgets to connect to the window
    :param title: Title to set to window, if renderToImage is false.
    :param fullscreen: Enable fullscreen window, if renderToImage is false.
    :param maximize: Maximize window, if renderToImage is false.
    :return: window if returnWindow is set to True, else None
    """

    width = _set_default_shorcut_value(width, 0)
    height = _set_default_shorcut_value(height, 0)
    intensityLevel = _set_default_shorcut_value(intensityLevel, -1)
    intensityWindow = _set_default_shorcut_value(intensityWindow, -1)
    segmentationColors = _set_default_shorcut_value(segmentationColors, LabelColors())
    segmentationBorderOpacity = _set_default_shorcut_value(segmentationBorderOpacity, -1)
    transferFunction = _set_default_shorcut_value(transferFunction, TransferFunction())

    if displayType == DisplayType.SLICER:
        window = fast.SlicerWindow.create(bgcolor=bgcolor, width=width, height=height)
        if image is not None:
            window.connectImage(image, intensityLevel, intensityWindow)

        if segmentation is not None:
            window.connectSegmentation(
                segmentation,
                colors=segmentationColors,
                opacity=segmentationOpacity,
                borderOpacity=segmentationBorderOpacity,
                borderRadius=segmentationBorderRadius
            )

    else:
        window = fast.SimpleWindow3D.create(bgcolor=bgcolor, width=width, height=height)
        if displayType == DisplayType.MAXIMUM_INTENSITY_PROJECTION:
            renderer = fast.MaximumIntensityProjection.create().connect(image)
            window.connect(renderer)
        elif displayType == DisplayType.ALPHA_BLENDING:
            renderer = fast.AlphaBlendingVolumeRenderer.create(transferFunction).connect(image)
            window.connect(renderer)

    if title: window.setTitle(title)
    if fullscreen: window.enableFullscreen()
    if maximize: window.enableMaximized()
    if timeout is not None: window.setTimeout(timeout)
    if widgets:
        if isinstance(widgets, dict):
            for pos, widget_list in widgets.items():
                window.connect(widget_list, pos)
        else:
            window.connect(widgets)
    if returnWindow:
        return window
    else:
        window.run()

